src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -1220,102 +1222,876 @@ private static float MinMaxCore<TMinMaxOperator>(ReadOnlySpan<float> x)
+            // in a way that allows us to have the minimum possible
+            // for small sizes
+
             ref float xRef = ref MemoryMarshal.GetReference(x);
             ref float dRef = ref MemoryMarshal.GetReference(destination);
-            int i = 0, oneVectorFromEnd;
+
+            nuint remainder = (uint)(x.Length);
 
 #if NET8_0_OR_GREATER
             if (Vector512.IsHardwareAccelerated)
             {
-                oneVectorFromEnd = x.Length - Vector512<float>.Count;
-                if (i <= oneVectorFromEnd)
+                if (remainder >= (uint)(Vector512<float>.Count))
                 {
-                    // Loop handling one vector at a time.
-                    do
+                    Vectorized512(ref xRef, ref dRef, remainder);
+                }
+                else
+                {
+                    // We have less than a vector and so we can only handle this as scalar. To do this
+                    // efficiently, we simply have a small jump table and fallthrough. So we get a simple
+                    // length check, single jump, and then linear execution.
+
+                    Vectorized512Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+#endif
+
+            if (Vector256.IsHardwareAccelerated)
+            {
+                if (remainder >= (uint)(Vector256<float>.Count))
+                {
+                    Vectorized256(ref xRef, ref dRef, remainder);
+                }
+                else
+                {
+                    // We have less than a vector and so we can only handle this as scalar. To do this
+                    // efficiently, we simply have a small jump table and fallthrough. So we get a simple
+                    // length check, single jump, and then linear execution.
+
+                    Vectorized256Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+
+            if (Vector128.IsHardwareAccelerated)
+            {
+                if (remainder >= (uint)(Vector128<float>.Count))
+                {
+                    Vectorized128(ref xRef, ref dRef, remainder);
+                }
+                else
+                {
+                    // We have less than a vector and so we can only handle this as scalar. To do this
+                    // efficiently, we simply have a small jump table and fallthrough. So we get a simple
+                    // length check, single jump, and then linear execution.
+
+                    Vectorized128Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+
+            // This is the software fallback when no acceleration is available
+            // It requires no branches to hit
+
+            SoftwareFallback(ref xRef, ref dRef, remainder);
+
+            static void SoftwareFallback(ref float xRef, ref float dRef, nuint length)
```

Comment:
Why separate SoftwareFallback out into its own method (which presumably isn't going to get inlined automatically)?

Reply:
If the dispatch method is "too large" or has "too many locals", the JIT will generate worse code.

In practice, only one of these branches will be "live" and we'll either have just the `SoftwareFallback` or just a given `Vectorized###` and `Vectorized###Small`.

We could likely mark this one as `AggressiveInlining`, but it's probably more profitable to mark `InvokeSpanIntoSpan` as such so the argument validation is inlined into the root entry point and then the dispatch to the core handler (`SoftwareFallback`, `Vectorized128`, etc) is the call instead

Reply:
> If the dispatch method is "too large" or has "too many locals", the JIT will generate worse code.

Shouldn't that only matter after it's already trimmed away all but the relevant scalar/128/256/512 path?

>  but it's probably more profitable to mark InvokeSpanIntoSpan as such so the argument validation is inlined into the root entry point

That would likely put all of that at the user's call site, since most of the calls in our methods to InvokeSpanIntoSpan are just the call.  That doesn't seem particularly desirable.

Reply:
> Shouldn't that only matter after it's already trimmed away all but the relevant scalar/128/256/512 path?

Depends on a few factors and the JIT is getting better about it, but it still has issues at times.

> That would likely put all of that at the user's call site, since most of the calls in our methods to InvokeSpanIntoSpan are just the call. That doesn't seem particularly desirable.

It depends on where the inlining happens. I doubt we'd end up with the `InvokeSpanIntoSpan(...)` inlined to the user call site. Rather instead we'd probably have it inlined into `Abs(...)` and the user call go to that instead.

But, it probably doesn't matter too much and we can mark `SoftwareFallback` and `Vector###Small` as `AggressiveInlining` if you prefer that. It's just beneficial to not have them in the main body so the main dispatch algorithm can participate in normal inlining heuristics around constant inputs and the like.

Reply:
> But, it probably doesn't matter too much and we can mark SoftwareFallback and Vector###Small as AggressiveInlining if you prefer that.

I don't currently see a downside to doing so (since that's the only call site, and if that code path is compiled in, you're going to call the method), and I can imagine it could save a couple nanoseconds, which could be measurable for small inputs.  So my default would be to mark it as such, unless perf tests show that it actually doesn't matter.

Reply:
Marked with AggressiveInlining.